test/modules/EdgeHubRestartTester/src/Program.cs

```diff
@@ -0,0 +1,161 @@
+                        sendMessageTask =
+                            new MessageEdgeHubConnector(
+                                msgModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                cts.Token,
+                                restartTime,
+                                restartStatus,
+                                restartCount,
+                                Logger)
+                            .ConnectAsync();
+                    }
+
+                    // Setup Direct Method Task
+                    Task directMethodTask = Task.CompletedTask;
+                    if (Settings.Current.DirectMethodEnable)
+                    {
+                        directMethodTask =
+                            new DirectMethodEdgeHubConnector(
+                                dmModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                cts.Token,
+                                restartTime,
+                                restartStatus,
+                                restartCount,
+                                Logger)
+                            .ConnectAsync();
+                    }
+
+                    // Wait for the two task to be done before do a restart
+                    await Task.WhenAll(new[] { sendMessageTask, directMethodTask });
+
+                    // Wait until the specified restart period to do another restart
+                    await Task.Delay((int)(eachTestExpirationTime - DateTime.UtcNow).TotalMilliseconds, cts.Token);
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Exception caught: {e}");
+                throw;
+            }
+            finally
+            {
+                iotHubServiceClient?.Dispose();
+                dmModuleClient?.Dispose();
+                msgModuleClient?.Dispose();
+            }
+
+            await cts.Token.WhenCanceled();
+            completed.Set();
+            handler.ForEach(h => GC.KeepAlive(h));
+            Logger.LogInformation("EdgeHubRestartTester Main() finished.");
+            return 0;
+        }
+
+        static async Task<Tuple<DateTime, HttpStatusCode>> RestartModules(
+            ServiceClient iotHubServiceClient)
+        {
+            CloudToDeviceMethod c2dMethod = new CloudToDeviceMethod("RestartModule");
+            string payloadSchema = "{{ \"SchemaVersion\": \"1.0\", \"Id\": \"{0}\" }}";
+            string payload = string.Format(payloadSchema, "edgeHub");
+            Logger.LogInformation("RestartModule Method Payload: {0}", payload);
+            c2dMethod.SetPayloadJson(payload);
+
+            try
+            {
+                CloudToDeviceMethodResult response = await iotHubServiceClient.InvokeDeviceMethodAsync(Settings.Current.DeviceId, "$edgeAgent", c2dMethod);
```

Comment:
I remembered that Varun said that you should even just restart the module by docker command.   @varunpuranik, please confirm.

Reply:
The offline scenario is not handled in this PR. Add `TODO` comments